src/walkers.rs

```diff
@@ -0,0 +1,172 @@
+    pub fn reset(&mut self) {
+        self.visited.clear();
+    }
+}
+
+impl VisitBlock for PreOrderCFG {
+    fn blocks_visit_order(
+        &mut self,
+        ctx: &Context,
+        cur_block: Ptr<BasicBlock>,
+    ) -> Vec<VisitAction<Ptr<BasicBlock>>> {
+        if self.visited.contains(&cur_block) {
+            return vec![];
+        }
+        let mut blocks = vec![VisitAction::Process];
+        blocks.extend(
+            cur_block
+                .deref(ctx)
+                .successors(ctx)
+                .into_iter()
+                .map(VisitAction::Visit),
+        );
+        blocks
+    }
+
+    fn operations_visit_order(
+        &mut self,
+        ctx: &Context,
+        cur_block: Ptr<BasicBlock>,
+    ) -> Vec<Ptr<Operation>> {
+        cur_block.deref(ctx).iter(ctx).collect()
+    }
+}
+
+/// At every node, process the node first,
+/// and then its successor or children.
+pub struct PreOrderVisitor {
+    cfg: PreOrderCFG,
+}
+
+impl Default for PreOrderVisitor {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl PreOrderVisitor {
+    pub fn new() -> Self {
+        PreOrderVisitor {
+            cfg: PreOrderCFG::default(),
+        }
+    }
+}
+
+// Delegate to self.cfg
+impl VisitBlock for PreOrderVisitor {
+    fn blocks_visit_order(
+        &mut self,
+        ctx: &Context,
+        cur_block: Ptr<BasicBlock>,
+    ) -> Vec<VisitAction<Ptr<BasicBlock>>> {
+        self.cfg.blocks_visit_order(ctx, cur_block)
+    }
+
+    fn operations_visit_order(
+        &mut self,
+        ctx: &Context,
+        cur_block: Ptr<BasicBlock>,
+    ) -> Vec<Ptr<Operation>> {
+        self.cfg.operations_visit_order(ctx, cur_block)
+    }
+}
+
+impl VisitOperation for PreOrderVisitor {
+    fn region_visit_order(
+        &mut self,
+        ctx: &Context,
+        cur_operation: Ptr<Operation>,
+    ) -> Vec<Ptr<Region>> {
+        cur_operation.deref(ctx).regions().collect()
+    }
+}
+
+impl VisitRegion for PreOrderVisitor {
+    fn blocks_visit_order(
+        &mut self,
+        ctx: &Context,
+        cur_region: Ptr<Region>,
+    ) -> Vec<Ptr<BasicBlock>> {
+        self.cfg.reset();
+        cur_region
+            .deref(ctx)
+            .get_head()
+            .map(|entry| vec![entry])
+            .unwrap_or_default()
+    }
+}
+
+/// A callback to process an [Operation], with the walker at `State`.
+pub type OpCallback<State> = fn(&mut State, &mut Context, Ptr<Operation>) -> Result<()>;
+/// A callback to process a [BasicBlock], with the walker at `State`.
+pub type BlockCallback<State> = fn(&mut State, &mut Context, Ptr<BasicBlock>) -> Result<()>;
+/// A callback to process an [Region], with the walker at `State`.
+pub type RegionCallback<State> = fn(&mut State, &mut Context, Ptr<Region>) -> Result<()>;
+
+pub struct PreOrderProcessor<State> {
+    pub state: State,
+    op_callback: OpCallback<State>,
+    block_callback: BlockCallback<State>,
+    region_callback: RegionCallback<State>,
+}
```

Comment:
I don't think this type is very helpful. Especially if all you want to operate on only one of the IR types. For that purpose I have implemented the generic `Callback` struct, which I did based on the callback based walkers that MLIR provides. If users want to operate on more then one IR type at a time, maybe it is easier to require them to implement the trait in that case?